### PR TITLE
Added Docker setup to build wasp-os.

### PR DESCRIPTION
**Pull request by shuhaowu**
This should make it a bit easier for people to contribute. Also fixed the documentations on how to setup on Debian, as certain recommended packages don't exist.

Signed-off-by: Shuhao Wu <shuhao@shuhaowu.com>

### CODE REVIEW

File: docs/install.rst [hunk 1]
```diff
@@ -50,6 +53,24 @@ tested using the `GNU-RM toolchain
...
+
+To build via docker, simply invoke `docker/shell` from the root directory after
+having docker installed. This will build the docker image and also drop you
+into a shell with wasp os's source code shared into the container at
+``/home/user/wasp-os``. All ``make`` commands should be usable from this shell,
+including ``make sim`` and ``make check``.
+
+.. note::
+
+    If you want to use the Docker-based setup, it is assumed that you're using
```
Comment (daniel-thompson) [inline docs/install.rst:67]: I think this also needs to mention that docker does not support the BLE tools (wasptool, ota-dfu, etc) so not all functionality is available via docker.

File: docs/install.rst [hunk 2]
```diff
@@ -50,6 +53,24 @@ tested using the `GNU-RM toolchain
...
+    and x86 machine on Linux, running Xorg. Other setup may require some
+    patching for now.
+
+Actually building wasp-os
+~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (daniel-thompson) [inline docs/install.rst:72]: Let's drop the actually from this header.

**Review by daniel-thompson — CHANGES_REQUESTED** [review]
Thanks for the PR. Most of the reviiew comments are about the doucmentation but there is on comment about the dockerfile too.

File: docs/install.rst [hunk 3]
```diff
@@ -21,8 +24,8 @@ following commands:
     sudo apt install \
       wget git build-essential libsdl2-2.0.0 python3-click python3-numpy \
       python3-pexpect python3-pil python3-pip python3-pydbus \
-      python-pygobject python3-serial unzip
-    pip3 install --user cbor pysdl2
+      python3-serial unzip
+    pip3 install --user cbor pysdl2 pygobject cryptography
```
[the remark this answers is not in the document]
Comment (shuhaowu) [reply docs/install.rst:28]: I needed cryptography. Without it, things like reloader-mcuboot.zip wouldn't build correctly as something complains about the inability to import `cryptography`.

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+RUN set -xe; \
+    export DEBIAN_FRONTEND=noninteractive; \
+    apt-get update; \
+    apt-get install -y \
+      wget git build-essential pkg-config libsdl2-2.0.0 python3-click \
+      python3-numpy python3-pexpect python3-pil python3-pip python3-pydbus \
+      libcairo2-dev python3-serial unzip python3-sphinx graphviz \
+      python3-recommonmark python3-pytest \
+    ; \
+    pip3 install cbor pysdl2 pygobject cryptography;
```
[the remark this answers is not in the document]
Comment (shuhaowu) [reply docker/Dockerfile:15]: This image is based on the Ubuntu official image and therefore we need to install all dependencies either manually. I don't think that we should use a Github image, as that would likely be pretty bloated with unnecessary dependencies and therefore would use extra space when you pull the image down.

File: docs/install.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (shuhaowu) [reply docs/install.rst]: Here's the trace

```
Zip created at build-pinetime/bootloader-daflasher.zip
make -C reloader/ BOARD=pinetime
make[1]: Entering directory '/home/user/wasp-os/reloader'
arm-none-eabi-gcc -mcpu=cortex-m4 -mthumb -g3 -Os -ffunction-sections -fdata-sections -DNRF52832_XXAA -Ilib/CMSIS/CMSIS/Core/Include -Ilib/nrfx -Ilib/nrfx/drivers/include -Ilib/nrfx/mdk -Ilib/nrfx/hal -Isrc/boards/pinetime -Isrc -Wl,-Map=build-pinetime/reloader.map -Lsrc -Llib/nrfx/mdk -Tnrf52832_xxaa.ld build-pinetime/main.o build-pinetime/blit.o build-pinetime/flash.o build-pinetime/spi.o build-pinetime/spinor.o build-pinetime/st7789.o build-pinetime/wdt.o build-pinetime/nrfx_nvmc.o build-pinetime/gcc_startup_nrf52.o build-pinetime/system_nrf52.o -o build-pinetime/reloader.elf
arm-none-eabi-objcopy -O ihex build-pinetime/reloader.elf build-pinetime/reloader.hex
python3 -m nordicsemi dfu genpkg --dev-type 0x0052 --application build-pinetime/reloader.hex build-pinetime/reloader.zip
Failed to import ecdsa, cannot do signing
Zip created at build-pinetime/reloader.zip
arm-none-eabi-gcc -mcpu=cortex-m4 -mthumb -g3 -Os -ffunction-sections -fdata-sections -DNRF52832_XXAA -Ilib/CMSIS/CMSIS/Core/Include -Ilib/nrfx -Ilib/nrfx/drivers/include -Ilib/nrfx/mdk -Ilib/nrfx/hal -Isrc/boards/pinetime -Isrc -Wl,-Map=build-pinetime/reloader.map -Lsrc -Llib/nrfx/mdk -Tnrf52832_xxaa_mcuboot.ld build-pinetime/main.o build-pinetime/blit.o build-pinetime/flash.o build-pinetime/spi.o build-pinetime/spinor.o build-pinetime/st7789.o build-pinetime/wdt.o build-pinetime/nrfx_nvmc.o build-pinetime/gcc_startup_nrf52.o build-pinetime/system_nrf52.o -o build-pinetime/reloader-mcuboot.elf
arm-none-eabi-objcopy -O binary build-pinetime/reloader-mcuboot.elf build-pinetime/reloader-mcuboot.bin.tmp
lib/mcuboot/scripts/imgtool.py create \
	--align 4 --version 1.0.0 \
	--header-size 32 --slot-size 475136 --pad-header build-pinetime/reloader-mcuboot.bin.tmp build-pinetime/reloader-mcuboot.bin
Traceback (most recent call last):
  File "lib/mcuboot/scripts/imgtool.py", line 17, in <module>
    from imgtool import main
  File "/home/user/wasp-os/reloader/lib/mcuboot/scripts/imgtool/main.py", line 21, in <module>
    import imgtool.keys as keys
  File "/home/user/wasp-os/reloader/lib/mcuboot/scripts/imgtool/keys/__init__.py", line 19, in <module>
    from cryptography.hazmat.backends import default_backend
ModuleNotFoundError: No module named 'cryptography'
make[1]: *** [Makefile:56: build-pinetime/reloader-mcuboot.bin] Error 1
make[1]: Leaving directory '/home/user/wasp-os/reloader'
make: *** [Makefile:51: reloader] Error 2
```

After that errors:

```
$ ls -l build-pinetime/
total 632
-rw-r--r-- 1 user user 170609 Feb  4 01:06 bootloader-daflasher.zip
-rw-r--r-- 1 user user 474700 Feb  4 01:06 bootloader.hex

```

File: docker/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (daniel-thompson) [reply docker/Dockerfile]: I didn't mean to suggest we should adopt the github image. Simply that I'd like to docker image to be more closely aligned with the CI flow (e.g. same choices of pip3 versus dpkg) to make it easier to maintain.

File: docs/install.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (daniel-thompson) [reply docs/install.rst]: Sorry... got confused between two different issues.

cryptography is only needed for pinetime builds... but its too complicated to spell out the differences in the install docs so it's correct to add it as a prerequisite.

**Comment by daniel-thompson** [comment]
Merged. Thanks.

**Pull request closed without merging**